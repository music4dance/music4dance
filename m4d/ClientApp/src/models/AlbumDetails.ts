import { jsonMember, jsonObject } from "typedjson";
import { PurchaseEncoded } from "./Purchase";

@jsonObject
export class AlbumDetails {
  @jsonMember(Number) public index?: number;
  @jsonMember(String) public name?: string;
  @jsonMember(Number) public track?: number;
  @jsonMember(String) public publisher?: string;
  @jsonMember(PurchaseEncoded) public purchase!: PurchaseEncoded;

  public constructor(init?: Partial<AlbumDetails>) {
    Object.assign(this, init);
    if (!this.purchase) {
      this.purchase = new PurchaseEncoded();
    }
  }
}
