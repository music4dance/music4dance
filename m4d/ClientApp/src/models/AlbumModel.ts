import { jsonMember, jsonObject } from "typedjson";
import { SongListModel } from "./SongListModel";

@jsonObject
export class AlbumModel extends SongListModel {
  @jsonMember(String) public title!: string;
  @jsonMember(String) public artist?: string;
}
