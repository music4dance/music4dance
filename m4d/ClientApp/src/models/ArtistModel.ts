import { jsonMember, jsonObject } from "typedjson";
import { SongListModel } from "./SongListModel";

@jsonObject
export class ArtistModel extends SongListModel {
  @jsonMember(String) public artist!: string;
}
