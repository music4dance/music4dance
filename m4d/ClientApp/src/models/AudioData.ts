import { jsonMember, jsonObject } from "typedjson";

@jsonObject
export class AudioData {
  @jsonMember(Number) public beatsPerMeasure?: number;
  @jsonMember(Number) public beatsPerMinute?: number;
  @jsonMember(Number) public danceability?: number;
  @jsonMember(Number) public energy?: number;
  @jsonMember(Number) public valence?: number;
}
