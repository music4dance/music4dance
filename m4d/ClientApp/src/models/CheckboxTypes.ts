import { wordsToKebab } from "@/helpers/StringHelpers";
import type { CheckboxOption, CheckboxValue } from "bootstrap-vue-next";

// INTO-TODO: See if we can merge ListOption into these

export interface Comparable {
  equals(that?: unknown): boolean;
}

export function isComparable(value: Comparable | unknown): value is Comparable {
  return (value as Comparable).equals !== undefined;
}

export function textFromOptions(options: CheckboxOption[]): string[] {
  return options.map(({ text }) => text);
}

export function valuesFromOptions(options: CheckboxOption[]): CheckboxValue[] {
  return options.map(({ value }) => value);
}

export function textFromValues(values: CheckboxValue[], all: CheckboxOption[]): string[] {
  return all.filter((o) => values.includes(o.value)).map((o) => o.text);
}

export function optionsFromText(text: string[]): CheckboxOption[] {
  return text.map((t) => ({ text: t, value: wordsToKebab(t) }));
}
