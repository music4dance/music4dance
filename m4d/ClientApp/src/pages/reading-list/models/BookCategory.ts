import type { Book } from "./BookModel";
import { type BookSelector, BookType } from "./BookType";

export class BookCategory {
  public books: Book[];

  constructor(
    public type: BookType,
    public title: string,
    public description: string,
    books: Book[],
  ) {
    this.books = BookCategory.extractCategory(books, type);
  }

  private static extractCategory(books: Book[], type: BookType): Book[] {
    return books.filter((b) => BookCategory.selectorContains(b.type, type));
  }

  private static selectorContains(selector: BookSelector, type: BookType): boolean {
    if (Array.isArray(selector)) {
      const rg = selector as BookType[];
      const idx = rg.findIndex((t) => t === type);
      return idx != -1;
    } else {
      const stype = selector as BookType;
      return stype == type;
    }
  }
}
