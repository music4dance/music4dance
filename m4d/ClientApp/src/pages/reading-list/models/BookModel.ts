import { type BookSelector } from "./BookType";

export interface Book {
  type: BookSelector;
  image: string;
  title: string;
  subtitle?: string;
  author: string;
  kindle?: string;
  paperback?: string;
  hardcover?: string;
  audible?: string;
  others?: Link[];
  review?: string;
  notes?: string;
  website?: Link;
}

export interface Link {
  text: string;
  ref: string;
}
