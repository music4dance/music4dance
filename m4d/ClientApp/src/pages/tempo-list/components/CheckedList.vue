<script setup lang="ts">
import { computed } from "vue";
import { type CheckboxValue, type CheckboxOption } from "bootstrap-vue-next";
import { valuesFromOptions, isComparable } from "@/models/CheckboxTypes";
import { wordsToKebab } from "@/helpers/StringHelpers";

const model = defineModel<CheckboxValue[]>({ required: true });
const props = defineProps<{
  type: string;
  options: CheckboxOption[];
}>();

const title = computed(() => {
  if (allSelected.value) {
    return allDescription.value;
  } else if (indeterminate.value) {
    if (model.value.length === 1) {
      const v = model.value[0];
      if (isComparable(v)) {
        return props.options.find((e) => v.equals(e.value))!.text;
      } else {
        return props.options.find((e) => e.value === model.value[0])!.text;
      }
    } else {
      return model.value.length + " " + pluralType.value;
    }
  } else {
    return noDescription.value;
  }
});

const computedId = wordsToKebab(props.type);

const allSelected = computed(() => {
  return model.value.length === props.options.length;
});

const indeterminate = computed(() => {
  return model.value.length > 0 && !allSelected.value;
});

const allDescription = computed(() => {
  return "All " + pluralType.value;
});

const noDescription = computed(() => {
  return "No " + pluralType.value;
});

const pluralType = computed(() => {
  return props.type + "s";
});

const allValues = computed(() => {
  return valuesFromOptions(props.options);
});

function toggleAll(checked: CheckboxValue | CheckboxValue[]): any {
  model.value = checked ? allValues.value : [];
}

// Do we need this in order to emit???
// watch(model, (newVal: CheckboxValue[]) => {
//   if (newVal.length === 0) {
//     indeterminate.value = false;
//     allSelected.value = false;
//   } else if (newVal.length === props.options.length) {
//     indeterminate.value = false;
//     allSelected.value = true;
//   } else {
//     indeterminate.value = true;
//     allSelected.value = false;
//   }
//   //emit("change", newVal);
// });
</script>

<template>
  <BDropdown
    id="dropdown-form"
    ref="dropdown"
    :text="title"
    variant="primary"
    style="margin-bottom: 8px"
    auto-close="outside"
  >
    <BDropdownForm>
      <BFormCheckbox
        :id="computedId + '-all'"
        :model-value="allSelected"
        :indeterminate="indeterminate"
        aria-describedby="selected"
        aria-controls="selected"
        data-test="select-all"
        @update:model-value="toggleAll"
      >
        {{ allSelected ? "Un-select All" : "Select All" }}
      </BFormCheckbox>
      <hr />
      <BFormGroup>
        <BFormCheckboxGroup
          :id="computedId + '-group'"
          v-model="model"
          :options="options"
          name="temp"
          stacked
          data-test="checkbox-group"
        />
      </BFormGroup>
    </BDropdownForm>
  </BDropdown>
</template>
