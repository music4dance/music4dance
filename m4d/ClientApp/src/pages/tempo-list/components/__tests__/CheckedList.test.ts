import { describe, expect, test } from "vitest";
import { VueWrapper, mount } from "@vue/test-utils";
import CheckedList from "../CheckedList.vue";
import { optionsFromText, textFromOptions, valuesFromOptions } from "@/models/CheckboxTypes";
import { type CheckboxValue } from "bootstrap-vue-next";

const textA = ["First Option", "Second Option", "Third Option"];
const optionsA = optionsFromText(textA);
const valuesA = valuesFromOptions(optionsA);

function getTextForParent(wrapper: VueWrapper, selector: string): string {
  const all = wrapper.find(selector);
  expect(all).toBeDefined();
  const allParent = all.element.parentElement;
  expect(allParent).toBeDefined();
  return allParent!.textContent ?? "";
}

describe("CheckedList.vue", () => {
  test("verify the options setup", () => {
    expect(optionsA).toBeDefined();
    expect(optionsA.length).toBe(3);
    expect(textFromOptions(optionsA)).toEqual(textA);
    expect(valuesA).toEqual(["first-option", "second-option", "third-option"]);
  });

  test("renders a simple checked list with all values checked", () => {
    const wrapper = mount(CheckedList, {
      propsData: { type: "test", options: optionsA, modelValue: valuesA },
    });
    expect(getTextForParent(wrapper, "input[data-test='select-all']")).toBe("Un-select All");
  });

  test("renders a simple checked list with no values checked", () => {
    const wrapper = mount(CheckedList, {
      propsData: { type: "test", options: optionsA, modelValue: [] },
    });
    expect(getTextForParent(wrapper, "input[data-test='select-all']")).toBe("Select All");
    const items = wrapper.findAll("#test-group input");
    expect(items).toBeDefined();
    expect(items.length).toBe(3);
    items.forEach((item) => {
      expect(item.attributes("checked")).toBeFalsy();
    });
  });

  test.skip("handle checking a single item", async () => {
    // TODO: Get this test working - I suspect I'm doing something wrong with
    //  handing of models/events especially as they work with respect to checkbox controls
    const wrapper = mount(CheckedList, {
      attachTo: document.body,
      propsData: { type: "test", options: optionsA, modelValue: [] },
      "onUpdate:model-value": (e: CheckboxValue[]) => {
        //console.log("onUpdate:modelValue", JSON.stringify(e));
        wrapper.setProps({ modelValue: e });
      },
    });

    //const all = wrapper.get("input[data-test='select-all']");
    //console.log("Initial classes:", all.classes().join(", "));

    const items = wrapper.findAll("#test-group input");
    expect(items[0].attributes("checked")).toBeFalsy();
    await items[0].trigger("click");
    //const emits = wrapper.emitted("update:modelValue");
    //console.log("Emits:", JSON.stringify(emits));
    expect(items[0].attributes("checked")).toBeTruthy();

    expect(getTextForParent(wrapper, "input[data-test='select-all']")).toBe("Select All");
    //const items = wrapper.findAll("#test-group input");
  });

  test.skip("renders a simple checked list with all values checked (snapshot)", () => {
    const wrapper = mount(CheckedList, {
      propsData: { type: "test", options: optionsA, modelValue: valuesA },
    });
    expect(wrapper.html()).toMatchSnapshot();
    const items = wrapper.findAll("#test-group input");
    expect(items).toBeDefined();
    expect(items.length).toBe(3);
    items.forEach((item) => {
      expect(item.attributes("checked")).toBeTruthy();
    });
  });

  test.skip("renders a simple checked list with no values checked (snapshot)", () => {
    const wrapper = mount(CheckedList, {
      propsData: { type: "test", options: optionsA, modelValue: [] },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
