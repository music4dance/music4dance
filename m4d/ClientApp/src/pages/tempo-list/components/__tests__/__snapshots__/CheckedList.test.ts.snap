// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CheckedList.vue > renders a simple checked list with all values checked (snapshot) 1`] = `
"<div class=\\"btn-group\\" style=\\"margin-bottom: 8px;\\"><button class=\\"btn btn-md btn-primary dropdown-toggle\\" type=\\"button\\" id=\\"dropdown-form\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\">All tests</button>
  <!---->
  <ul style=\\"position: absolute; left: 0px; top: 0px; display: none;\\" class=\\"dropdown-menu show\\" aria-labelledby=\\"dropdown-form\\" role=\\"menu\\">
    <li role=\\"presentation\\">
      <form class=\\"px-4 py-3\\">
        <div class=\\"form-check\\"><input id=\\"test-all\\" aria-describedby=\\"selected\\" aria-controls=\\"selected\\" data-test=\\"select-all\\" class=\\"form-check-input\\" type=\\"checkbox\\" value=\\"true\\"><label for=\\"test-all\\" class=\\"form-check-label\\">Un-select All</label></div>
        <hr>
        <fieldset class=\\"\\" id=\\"__BVID__843474___BV___\\">
          <!---->
          <div class=\\"\\">
            <div id=\\"test-group\\" role=\\"group\\" class=\\"btn-group-md bv-no-focus-ring\\" tabindex=\\"-1\\" data-test=\\"checkbox-group\\">
              <div class=\\"form-check\\"><input id=\\"__BVID__016026___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"first-option\\"><label for=\\"__BVID__016026___BV_form-check__\\" class=\\"form-check-label\\"><span>First Option</span></label></div>
              <div class=\\"form-check\\"><input id=\\"__BVID__092157___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"second-option\\"><label for=\\"__BVID__092157___BV_form-check__\\" class=\\"form-check-label\\"><span>Second Option</span></label></div>
              <div class=\\"form-check\\"><input id=\\"__BVID__753942___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"third-option\\"><label for=\\"__BVID__753942___BV_form-check__\\" class=\\"form-check-label\\"><span>Third Option</span></label></div>
            </div>
            <!---->
            <!---->
            <!---->
          </div>
        </fieldset>
      </form>
    </li>
  </ul>
</div>"
`;

exports[`CheckedList.vue > renders a simple checked list with no values checked (snapshot) 1`] = `
"<div class=\\"btn-group\\" style=\\"margin-bottom: 8px;\\"><button class=\\"btn btn-md btn-primary dropdown-toggle\\" type=\\"button\\" id=\\"dropdown-form\\" aria-expanded=\\"false\\" aria-haspopup=\\"menu\\">No tests</button>
  <!---->
  <ul style=\\"position: absolute; left: 0px; top: 0px; display: none;\\" class=\\"dropdown-menu show\\" aria-labelledby=\\"dropdown-form\\" role=\\"menu\\">
    <li role=\\"presentation\\">
      <form class=\\"px-4 py-3\\">
        <div class=\\"form-check\\"><input id=\\"test-all\\" aria-describedby=\\"selected\\" aria-controls=\\"selected\\" data-test=\\"select-all\\" class=\\"form-check-input\\" type=\\"checkbox\\" value=\\"true\\"><label for=\\"test-all\\" class=\\"form-check-label\\">Select All</label></div>
        <hr>
        <fieldset class=\\"\\" id=\\"__BVID__373777___BV___\\">
          <!---->
          <div class=\\"\\">
            <div id=\\"test-group\\" role=\\"group\\" class=\\"btn-group-md bv-no-focus-ring\\" tabindex=\\"-1\\" data-test=\\"checkbox-group\\">
              <div class=\\"form-check\\"><input id=\\"__BVID__487684___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"first-option\\"><label for=\\"__BVID__487684___BV_form-check__\\" class=\\"form-check-label\\"><span>First Option</span></label></div>
              <div class=\\"form-check\\"><input id=\\"__BVID__392589___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"second-option\\"><label for=\\"__BVID__392589___BV_form-check__\\" class=\\"form-check-label\\"><span>Second Option</span></label></div>
              <div class=\\"form-check\\"><input id=\\"__BVID__672126___BV_form-check__\\" class=\\"form-check-input\\" type=\\"checkbox\\" name=\\"temp\\" value=\\"third-option\\"><label for=\\"__BVID__672126___BV_form-check__\\" class=\\"form-check-label\\"><span>Third Option</span></label></div>
            </div>
            <!---->
            <!---->
            <!---->
          </div>
        </fieldset>
      </form>
    </li>
  </ul>
</div>"
`;
